pygmt/src/grdhisteq.py

```
@@ -0,0 +1,369 @@
+
+            - pandas.DataFrame if ``outfile`` is True or None
+            - None if ``outfile`` is a str (file output is stored in
+              ``outfile``)
+
+        Example
+        -------
+        >>> import pygmt  # doctest: +SKIP
+        >>> # Load a grid of @earth_relief_30m data, with an x-range of 10 to
+        >>> # 30, and a y-range of 15 to 25
+        >>> grid = pygmt.datasets.load_earth_relief(
+        ...     resolution="30m", region=[10, 30, 15, 25]
+        ... )  # doctest: +SKIP
+        >>> # Find elevation intervals that splits the data range into 5
+        >>> # divisions, each of which have an equal area in the original grid.
+        >>> bins = pygmt.grdhisteq.compute_bins(
+        ...     grid=grid, divisions=5
+        ... )  # doctest: +SKIP
+        >>> print(bins)  # doctest: +SKIP
+                start    stop
+        bin_id
+        0       179.0   397.5
+        1       397.5   475.5
+        2       475.5   573.5
+        3       573.5   710.5
+        4       710.5  2103.0
+
+        See Also
+        -------
+        :meth:`pygmt.grd2cpt`
+
+        Notes
+        -----
+        This method does a weighted histogram equalization for geographic
+        grids to account for node area varying with latitude.
+        """
+        # Return a pandas.DataFrame if ``outfile`` is not set
+        if output_type not in ["numpy", "pandas", "file"]:
+            raise GMTInvalidInput(
+                "Must specify 'output_type' either as 'numpy', 'pandas' or 'file'."
+            )
+
+        if isinstance(outfile, str) and output_type != "file":
+            msg = (
+                f"Changing 'output_type' from '{output_type}' to 'file' "
+                "since 'outfile' parameter is set. Please use output_type='file' "
+                "to silence this warning."
+            )
+            warnings.warn(message=msg, category=RuntimeWarning, stacklevel=2)
+            output_type = "file"
```

Perhaps we should raise an exception here rather than give a warning? Sometimes people just ignore warnings.

---

There was a bit of a discussion about this following this comment https://github.com/GenericMappingTools/pygmt/pull/1284#pullrequestreview-737690502. See https://github.com/GenericMappingTools/pygmt/pull/1284#discussion_r698128719 for a specific discussion about raising a warning versus exception. I think we should follow that lead, although we could move this to a helper function to make it more clear that the logic is used for multiple modules.